lms/services/assignment.py

```diff
@@ -10,55 +15,206 @@ def __init__(self, db):
             )
 
         # Clear the cache (@lru_cache) on self.get because we've changed the
         # contents of the DB. (Python's @lru_cache doesn't have a way to remove
         # just one key from the cache, you have to clear the entire cache.)
         self.get.cache_clear()
 
+        return assignment
+
+    def _merge_canvas_assignments(self, assignments):
+        # We stored a canvas file assignment before (storing its resource_link_id)
+        # we later configured it (storing its ext_lti_assignment_id) in a new row
+        # and now we are launching it, we want to merge those two assignments
+        old_assignment = (
+            assignments[0] if assignments[0].resource_link_id else assignments[1]
+        )
+        assignment = (
+            assignments[0] if assignments[0].ext_lti_assignment_id else assignments[1]
+        )
+        self._db.delete(old_assignment)
+        self._db.flush()
+        return assignment
+
+    def _get_by_resource_link_id(self, tool_consumer_instance_guid, resource_link_id):
+        return (
+            self._db.query(Assignment)
+            .filter_by(
+                tool_consumer_instance_guid=tool_consumer_instance_guid,
+                resource_link_id=resource_link_id,
+            )
+            .one()
+        )
+
+    def _get_by_ext_lti_assignment(
+        self, tool_consumer_instance_guid, ext_lti_assignment_id
+    ):
+        return (
+            self._db.query(Assignment)
+            .filter_by(
+                tool_consumer_instance_guid=tool_consumer_instance_guid,
+                ext_lti_assignment_id=ext_lti_assignment_id,
+            )
+            .one()
+        )
+
+    def _get_for_canvas_launch(
+        self,
+        tool_consumer_instance_guid,
+        resource_link_id,
+        ext_lti_assignment_id,
+    ):
+        """
+        Get a canvas assignment by both resource_link_id and ext_lti_assignment_id.
+
+        Having both values means that we are most likely launching an assignment but
+        we might find these cases on the DB data:
+
+            - A row with both values.
+                An assignment that has been launched before.
+                Just get it an return it.
+
+            - A row with with just the `ext_lti_assignment_id`.
+                We have configured this assignment before but never launched it
+                so we don't yet know its `resource_link_id`.
+
+                Get it, fill the new value on `resource_link_id` and return it.
+
+            - Two rows, one with each of the ids an the other set to null.
+
+                In this case we stored the assignment in the past on launch with just the `resource_link_id`,
+                later, we re-configured it, creating a new row with just the `ext_lti_assignment_id`.
+
+                We need to get both and merge them into one.
+                We'll keep the one with `ext_lti_assignment_id` as is the newest one.
+        """
```

Comment:
```suggestion
        """Get a canvas assignment by both resource_link_id and ext_lti_assignment_id."""
```

I think we should document the different cases with code comments rather than in the docstring (see other comment below). The specific cases are sort of an internal implementation detail so probably don't belong in the docstring. And putting them in code comments gets them closer to the code they're documenting.